### ISSUE TITLE
test.py stopped running tests in cql-pytest subdirectories

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#10962]
By running `test/cql-pytest/run` manually, I discovered three failing tests, which I listed in https://github.com/scylladb/scylla/issues/10918.

Because we never saw these failures in Jenkins, I thought the first one was a miscompilation on my Fedora 36 machine. But the other two failures are simple error-string mismatches, and could not have possibly been missed.

Only now did I realize what happened: `test.py` simply stopped running these tests at some point, which is why nobody noticed that master now has three failing tests!

The tests in `test/cql-pytest` are not all in the top-level directory, there are also subdirectories, specifically the `cassandra_tests` subdirectory for translated Cassandra unit tests. One of the test files which fails but test.py misses is:

```
test/cql-pytest/cassandra_tests/validation/operations/select_test.py
```

If I try to run `./test.py select_test`, it can't find it!
If I run `./test.py  --mode=dev cql-pytest|cat` it confirms that misses all subdirectories - it just lists the files directly under `test/cql-pytest`.

This is a really bad regression in our testing, because it means that during the switch away from `test/cql-pytest/run` we simply stopped running a lot of our tests, and we didn't notice, and by now three tests are already failing (two are silly string mismatches, but one seems like a serious bug). @kostja can you please fix this?

**Comment by nyh** [comment]
By the way, we really need to fix this bug, but before we commit its fix we need first to either 1. commit a fix to the last failing test in issue #10918 or 2. revert the patch which caused it (see the bisection in that issue). If we commit a fix to **this** issue first, the test suite will start to fail.

**Comment by kostja** [comment]
I had no idea there were tests in subdirectories. Suites are not traversed recursively. Let's make it a separate suite by moving to an own dir - that would be a trivial fix.

**Comment by nyh** [comment]
@kostja let's **not** make it a separate suite.

Note that the directory `cassandra_tests` has its own subdirectories - I kept the same subdirectory names and file names as the original Cassandra unit test source files. I don't want to put them all in one big directory just because test.py cannot handle subdirectories.

To make a long story short, pytest collects tests from subdirectories, so your test.py's wrapper to pytest must do the same. Anything else is a bug. You wanted not to run "pytest" for everything at once and instead run "pytest filename" for each individual filename, then you need to be aware of pytest's rules of which filenames it collects. And that include subdirectories.

**Comment by nyh** [comment]
test.py is not a rule of nature - it's a tool that we write for our convenience.

In the present case, if test.py is intended as an easy way to run pytest, it should support whatever pytest supports. And in this case, since pytest supports subdirectories well, so should test.py when it runs a pytest-based suite. Yes, we can outlaw a certain pytest if there is a good reason to, but there is no good reason to outlaw the subdirectories feature which is well-documented (see https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#test-discovery).  Organizing tests in subdirectories is a convenient organization feature. In this case I wanted 1. to separate the translated tests from our own tests, to make it clearer which is which 2. to keep the same directory organization that Cassandra uses in their Java code.

I don't suggest that you change anything in how test.py finds suite.yaml. What I'm suggesting is once you have decided that the top-level test/cql-pytest directory is a pytest suite directory, then treat it like one - and this includes looking at subdirectories in the same way that pytest does.

**Comment by kostja** [comment]
There are assumptions that a file name inside a given suite is not a subpath on other levels. We derive log file name from test name, xml output file name. Recursion support can be added but it will be a cognitive burden.

**Comment by nyh** [comment]
@kostja there is no way around it - pytest supports subdirectories natively, this is well-documented, and already used in our tests, so you have to support it in test.py. It worked before your patch to stop running `test/cql-pytest/run.py`.
You can replace the slashes with dots or underscores or anything else in test names if the slash in test names bothers you.

**Comment by nyh** [comment]
The fixes to all cql-pytests failures (one real regression that was caught by a test, and two non-interesting message changes) are now in next, and test/cql-pytest/run, including all subdirectories, now passes.
So it's a good time to fix test.py to also run the pytests in subdirectories.

**Comment by kostja** [comment]
Working on it

**Comment by kostja** [comment]
So here's the issue. The pattern for cql-pytest is test_*.py. This is a convention for all Scylla unit tests - they have to be prefixed with test_. I don't know the reason for it, but now it's glorified in how we search for tests. The tests in cql-pytest subdirs don't match this pattern. Even when we start looking. 

If I kill the pattern, I have to look *into* the test like pytest does, to skip files like run.py and conftest.py. I think it's a bad idea. So in the end... shall I rename these tests?

**Comment by kostja** [comment]
@nyh

**Comment by nyh** [comment]
> So here's the issue. The pattern for cql-pytest is test_*.py. This is a convention for all Scylla unit tests - they have to be prefixed with test_. I don't know the reason for it, but now it's glorified in how we search for tests. The tests in cql-pytest subdirs don't match this pattern. Even when we start looking.

The pytest documentation https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html explains how pytest looks for tests:

1. Recurse into directories, unless they match norecursedirs
2. In those directories, search for **`test_*.py` or `*_test.py`** files

When I translated a Java test called SecondaryIndexTest.java, I called the resulting file secondary_index_test.py - because this was the most natural filename translation. I could have changed the order to test_secondary_index.py, but I didn't need to because it worked in pytest.

If for some reason it's very hard for test.py to allow this pytest convention (is it hard?) I guess we can rename the tests to reorder the word "test" in the file name. It wouldn't be a big disaster to change it. But personally, I think that since test.py is attempting to wrap pytest, it should make an effort to allow the same conventions that pytest allows.

> If I kill the pattern, I have to look _into_ the test like pytest does, to skip files like run.py and conftest.py. I think it's a bad idea. 

As I noted, pytest just adds a second globbing pattern. Files like run.py, conftest.py and util.py match neither pattern, so indeed will be skipped.

>So in the end... shall I rename these tests?

This is an option, but if the option to add a second globbing pattern isn't too difficult, I think it's better and easier.
It will also mean that pytest.py continues the tradition of allowing whatever pytest allows.

**Comment by nyh** [comment]
I just realized that we also have Boost tests like `test/boost/secondary_index_test.cc` which use this second globbing pattern (`*_test`), so I wonder how this was never a problem in the past.

**Comment by kostja** [comment]
All boost tests are _test.cc

**Comment by eliransin** [comment]
> just realized that we also have Boost tests like `test/boost/secondary_index_test.cc` which use this second globbing pattern (`*_test`), so I wonder how this was never a problem in the past.

@nyh I guess this is not relevant here since those are compiling to actual executables and not some pytest python 
classes/files. The real question is how does the pytest that wraps them is called.

> Thanks for explaining, it's trivial to add a second globbing pattern.

@kostja I'm unclear about the state of this issue, are we just in agreement or does this issue is actively handled?
IIUC every day that this regression is present, is another day that some tests (that are proven to find regressions in the past) are
not being run.
If this is a matter of bandwidth I guess @nyh and I can take ownership of fixing this (unless I am mistaken and there is no agreement about the way to solve this...)

**Comment by nyh** [comment]
The bug was fixed in a645bb3622351d55981c2c00159acd8f49a63913
Now you can do
```
./test.py --mode=dev cql-pytest
```
And it runs all cql-pytest tests, including those in cassandra_tests/ and its subdirectories.

The patch was missing a "Fixes" so we forgot to close this issue.
